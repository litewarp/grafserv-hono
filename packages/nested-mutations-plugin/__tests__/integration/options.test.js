const { graphql } = require('graphql');
const { withSchema } = require('../helpers');

test(
  'simple names, plural when one-to-many, singular in reverse',
  withSchema({
    setup: `
      create table p.parent (
        id serial primary key,
        name text not null
      );
      
      create table p.child (
        id serial primary key,
        parent_id integer,
        name text not null,
        constraint child_parent_fkey foreign key (parent_id)
          references p.parent (id)
      );   
    `,
    options: {
      graphileBuildOptions: {
        nestedMutationsSimpleFieldNames: true,
      },
    },
    test: async ({ schema, pgClient }) => {
      const query = `
        mutation {
          c1: createParent(
            input: {
              parent: {
                name: "test"
                children: {
                  create: [{
                    name: "test child"
                  }]
                }
              }
            }
          ) {
            parent {
              id
            }
          }

          c2: createChild(
            input: {
              child: {
                name: "child"
                parent: {
                  create: {
                    name: "child's parent"
                  }
                }
              }
            }
          ) {
            child {
              id
            }
          }
        }
      `;
      expect(schema).toMatchSnapshot();

      const result = await graphql(schema, query, null, { pgClient });
      expect(result).not.toHaveProperty('errors');
    },
  }),
);

test(
  'simple names, singular when one-to-one',
  withSchema({
    setup: `
      create table p.parent (
        id serial primary key,
        name text not null
      );
      
      create table p.child (
        parent_id serial primary key,
        name text not null,
        constraint child_parent_fkey foreign key (parent_id)
          references p.parent (id)
      );   
    `,
    options: {
      graphileBuildOptions: {
        nestedMutationsSimpleFieldNames: true,
      },
    },
    test: async ({ schema, pgClient }) => {
      const query = `
        mutation {
          c1: createParent(
            input: {
              parent: {
                name: "test"
                child: {
                  create: {
                    name: "test child"
                  }
                }
              }
            }
          ) {
            parent {
              id
            }
          }

          c2: createChild(
            input: {
              child: {
                name: "child"
                parent: {
                  create: {
                    name: "child's parent"
                  }
                }
              }
            }
          ) {
            child {
              parentId
            }
          }
        }
      `;
      expect(schema).toMatchSnapshot();

      const result = await graphql(schema, query, null, { pgClient });
      expect(result).not.toHaveProperty('errors');
    },
  }),
);

test(
  'deleteOthers is not avaialble when disabled',
  withSchema({
    setup: `
      create table p.parent (
        id serial primary key,
        name text not null
      );
      
      create table p.child (
        id serial primary key,
        parent_id integer,
        name text not null,
        constraint child_parent_fkey foreign key (parent_id)
          references p.parent (id)
      );
      insert into p.parent values(1, 'test');
      insert into p.child values(99, 1, 'test child');
    `,
    options: {
      graphileBuildOptions: {
        nestedMutationsDeleteOthers: false,
      },
    },
    test: async ({ schema, pgClient }) => {
      const query = `
        mutation {
          updateParentById(
            input: {
              id: 1
              parentPatch: {
                childrenUsingId: {
                  deleteOthers: true
                  create: [{
                    name: "test child 2"
                  }, {
                    name: "test child 3"
                  }]
                }
              }
            }
          ) {
            parent {
              id
              name
              childrenByParentId {
                nodes {
                  id
                  parentId
                  name
                }
              }
            }
          }
        }
      `;
      const result = await graphql(schema, query, null, { pgClient });
      expect(result).toHaveProperty('errors');
      expect(result.errors[0].message).toMatch(/"deleteOthers" is not defined/);
    },
  }),
);

test(
  'still plural when one-to-one if nestedMutationsOldUniqueFields is enabled',
  withSchema({
    setup: `
      create table p.parent (
        id serial primary key,
        name text not null
      );
      
      create table p.child (
        parent_id serial primary key,
        name text not null,
        constraint child_parent_fkey foreign key (parent_id)
          references p.parent (id)
      );   
    `,
    options: {
      graphileBuildOptions: {
        nestedMutationsOldUniqueFields: true,
      },
    },
    test: async ({ schema, pgClient }) => {
      const query = `
        mutation {
          createParent(
            input: {
              parent: {
                name: "test"
                childrenUsingId: {
                  create: {
                    name: "test child"
                  }
                }
              }
            }
          ) {
            parent {
              id
            }
          }
        }
      `;
      expect(schema).toMatchSnapshot();

      const result = await graphql(schema, query, null, { pgClient });
      expect(result).not.toHaveProperty('errors');
    },
  }),
);

// from https://github.com/mlipscombe/postgraphile-plugin-nested-mutations/issues/40
test(
  'id fields are renamed rowId when classicIds is enabled',
  withSchema({
    setup: `
      create table p.parent (
        id uuid primary key default public.uuid_generate_v4(),
        name text not null
      );
      create table p.child (
        id uuid primary key default public.uuid_generate_v4(),
        parent_id uuid references p.parent on delete set null,
        name text not null
      );
    `,
    options: {
      classicIds: true,
    },
    test: async ({ schema, pgClient }) => {
      const query = `
        mutation {
          createParent(
            input: {
              parent: {
                name: "test"
                childrenUsingRowId: {
                  create: {
                    name: "test child"
                  }
                }
              }
            }
          ) {
            parent {
              id
              rowId
              childrenByParentId {
                nodes {
                  id
                  rowId
                }
              }
            }
          }
        }
      `;
      expect(schema).toMatchSnapshot();

      const result = await graphql(schema, query, null, { pgClient });
      expect(result).not.toHaveProperty('errors');
    },
  }),
);
