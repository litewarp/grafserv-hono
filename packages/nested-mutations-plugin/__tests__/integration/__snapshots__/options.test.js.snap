// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`id fields are renamed rowId when classicIds is enabled 1`] = `
GraphQLSchema {
  "__validationErrors": Array [],
  "_directives": Array [
    "@include",
    "@skip",
    "@deprecated",
    "@specifiedBy",
  ],
  "_implementationsMap": Object {
    "Node": Object {
      "interfaces": Array [],
      "objects": Array [
        "Query",
        "Child",
        "Parent",
      ],
    },
  },
  "_mutationType": "Mutation",
  "_queryType": "Query",
  "_subTypeMap": Object {},
  "_subscriptionType": undefined,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "Child": "Child",
    "ChildChildPkeyConnect": "ChildChildPkeyConnect",
    "ChildChildPkeyDelete": "ChildChildPkeyDelete",
    "ChildCondition": "ChildCondition",
    "ChildInput": "ChildInput",
    "ChildNodeIdConnect": "ChildNodeIdConnect",
    "ChildNodeIdDelete": "ChildNodeIdDelete",
    "ChildOnChildForChildParentIdFkeyNodeIdUpdate": "ChildOnChildForChildParentIdFkeyNodeIdUpdate",
    "ChildOnChildForChildParentIdFkeyUsingChildPkeyUpdate": "ChildOnChildForChildParentIdFkeyUsingChildPkeyUpdate",
    "ChildParentIdFkeyChildCreateInput": "ChildParentIdFkeyChildCreateInput",
    "ChildParentIdFkeyInput": "ChildParentIdFkeyInput",
    "ChildParentIdFkeyInverseInput": "ChildParentIdFkeyInverseInput",
    "ChildParentIdFkeyParentCreateInput": "ChildParentIdFkeyParentCreateInput",
    "ChildPatch": "ChildPatch",
    "ChildrenConnection": "ChildrenConnection",
    "ChildrenEdge": "ChildrenEdge",
    "ChildrenOrderBy": "ChildrenOrderBy",
    "CreateChildInput": "CreateChildInput",
    "CreateChildPayload": "CreateChildPayload",
    "CreateParentInput": "CreateParentInput",
    "CreateParentPayload": "CreateParentPayload",
    "Cursor": "Cursor",
    "DeleteChildByRowIdInput": "DeleteChildByRowIdInput",
    "DeleteChildInput": "DeleteChildInput",
    "DeleteChildPayload": "DeleteChildPayload",
    "DeleteParentByRowIdInput": "DeleteParentByRowIdInput",
    "DeleteParentInput": "DeleteParentInput",
    "DeleteParentPayload": "DeleteParentPayload",
    "ID": "ID",
    "Int": "Int",
    "Mutation": "Mutation",
    "Node": "Node",
    "PageInfo": "PageInfo",
    "Parent": "Parent",
    "ParentCondition": "ParentCondition",
    "ParentInput": "ParentInput",
    "ParentNodeIdConnect": "ParentNodeIdConnect",
    "ParentNodeIdDelete": "ParentNodeIdDelete",
    "ParentOnChildForChildParentIdFkeyNodeIdUpdate": "ParentOnChildForChildParentIdFkeyNodeIdUpdate",
    "ParentOnChildForChildParentIdFkeyUsingParentPkeyUpdate": "ParentOnChildForChildParentIdFkeyUsingParentPkeyUpdate",
    "ParentParentPkeyConnect": "ParentParentPkeyConnect",
    "ParentParentPkeyDelete": "ParentParentPkeyDelete",
    "ParentPatch": "ParentPatch",
    "ParentsConnection": "ParentsConnection",
    "ParentsEdge": "ParentsEdge",
    "ParentsOrderBy": "ParentsOrderBy",
    "Query": "Query",
    "String": "String",
    "UUID": "UUID",
    "UpdateChildByRowIdInput": "UpdateChildByRowIdInput",
    "UpdateChildInput": "UpdateChildInput",
    "UpdateChildPayload": "UpdateChildPayload",
    "UpdateParentByRowIdInput": "UpdateParentByRowIdInput",
    "UpdateParentInput": "UpdateParentInput",
    "UpdateParentPayload": "UpdateParentPayload",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
    "updateChildOnChildForChildParentIdFkeyPatch": "updateChildOnChildForChildParentIdFkeyPatch",
    "updateParentOnChildForChildParentIdFkeyPatch": "updateParentOnChildForChildParentIdFkeyPatch",
  },
  "astNode": undefined,
  "description": undefined,
  "extensionASTNodes": undefined,
  "extensions": undefined,
}
`;

exports[`simple names, plural when one-to-many, singular in reverse 1`] = `
GraphQLSchema {
  "__validationErrors": Array [],
  "_directives": Array [
    "@include",
    "@skip",
    "@deprecated",
    "@specifiedBy",
  ],
  "_implementationsMap": Object {
    "Node": Object {
      "interfaces": Array [],
      "objects": Array [
        "Query",
        "Child",
        "Parent",
      ],
    },
  },
  "_mutationType": "Mutation",
  "_queryType": "Query",
  "_subTypeMap": Object {},
  "_subscriptionType": undefined,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "Child": "Child",
    "ChildChildPkeyConnect": "ChildChildPkeyConnect",
    "ChildChildPkeyDelete": "ChildChildPkeyDelete",
    "ChildCondition": "ChildCondition",
    "ChildInput": "ChildInput",
    "ChildNodeIdConnect": "ChildNodeIdConnect",
    "ChildNodeIdDelete": "ChildNodeIdDelete",
    "ChildOnChildForChildParentFkeyNodeIdUpdate": "ChildOnChildForChildParentFkeyNodeIdUpdate",
    "ChildOnChildForChildParentFkeyUsingChildPkeyUpdate": "ChildOnChildForChildParentFkeyUsingChildPkeyUpdate",
    "ChildParentFkeyChildCreateInput": "ChildParentFkeyChildCreateInput",
    "ChildParentFkeyInput": "ChildParentFkeyInput",
    "ChildParentFkeyInverseInput": "ChildParentFkeyInverseInput",
    "ChildParentFkeyParentCreateInput": "ChildParentFkeyParentCreateInput",
    "ChildPatch": "ChildPatch",
    "ChildrenConnection": "ChildrenConnection",
    "ChildrenEdge": "ChildrenEdge",
    "ChildrenOrderBy": "ChildrenOrderBy",
    "CreateChildInput": "CreateChildInput",
    "CreateChildPayload": "CreateChildPayload",
    "CreateParentInput": "CreateParentInput",
    "CreateParentPayload": "CreateParentPayload",
    "Cursor": "Cursor",
    "DeleteChildByIdInput": "DeleteChildByIdInput",
    "DeleteChildInput": "DeleteChildInput",
    "DeleteChildPayload": "DeleteChildPayload",
    "DeleteParentByIdInput": "DeleteParentByIdInput",
    "DeleteParentInput": "DeleteParentInput",
    "DeleteParentPayload": "DeleteParentPayload",
    "ID": "ID",
    "Int": "Int",
    "Mutation": "Mutation",
    "Node": "Node",
    "PageInfo": "PageInfo",
    "Parent": "Parent",
    "ParentCondition": "ParentCondition",
    "ParentInput": "ParentInput",
    "ParentNodeIdConnect": "ParentNodeIdConnect",
    "ParentNodeIdDelete": "ParentNodeIdDelete",
    "ParentOnChildForChildParentFkeyNodeIdUpdate": "ParentOnChildForChildParentFkeyNodeIdUpdate",
    "ParentOnChildForChildParentFkeyUsingParentPkeyUpdate": "ParentOnChildForChildParentFkeyUsingParentPkeyUpdate",
    "ParentParentPkeyConnect": "ParentParentPkeyConnect",
    "ParentParentPkeyDelete": "ParentParentPkeyDelete",
    "ParentPatch": "ParentPatch",
    "ParentsConnection": "ParentsConnection",
    "ParentsEdge": "ParentsEdge",
    "ParentsOrderBy": "ParentsOrderBy",
    "Query": "Query",
    "String": "String",
    "UpdateChildByIdInput": "UpdateChildByIdInput",
    "UpdateChildInput": "UpdateChildInput",
    "UpdateChildPayload": "UpdateChildPayload",
    "UpdateParentByIdInput": "UpdateParentByIdInput",
    "UpdateParentInput": "UpdateParentInput",
    "UpdateParentPayload": "UpdateParentPayload",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
    "updateChildOnChildForChildParentFkeyPatch": "updateChildOnChildForChildParentFkeyPatch",
    "updateParentOnChildForChildParentFkeyPatch": "updateParentOnChildForChildParentFkeyPatch",
  },
  "astNode": undefined,
  "description": undefined,
  "extensionASTNodes": undefined,
  "extensions": undefined,
}
`;

exports[`simple names, singular when one-to-one 1`] = `
GraphQLSchema {
  "__validationErrors": Array [],
  "_directives": Array [
    "@include",
    "@skip",
    "@deprecated",
    "@specifiedBy",
  ],
  "_implementationsMap": Object {
    "Node": Object {
      "interfaces": Array [],
      "objects": Array [
        "Query",
        "Child",
        "Parent",
      ],
    },
  },
  "_mutationType": "Mutation",
  "_queryType": "Query",
  "_subTypeMap": Object {},
  "_subscriptionType": undefined,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "Child": "Child",
    "ChildChildPkeyConnect": "ChildChildPkeyConnect",
    "ChildChildPkeyDelete": "ChildChildPkeyDelete",
    "ChildCondition": "ChildCondition",
    "ChildInput": "ChildInput",
    "ChildNodeIdConnect": "ChildNodeIdConnect",
    "ChildNodeIdDelete": "ChildNodeIdDelete",
    "ChildOnChildForChildParentFkeyNodeIdUpdate": "ChildOnChildForChildParentFkeyNodeIdUpdate",
    "ChildOnChildForChildParentFkeyUsingChildPkeyUpdate": "ChildOnChildForChildParentFkeyUsingChildPkeyUpdate",
    "ChildParentFkeyChildCreateInput": "ChildParentFkeyChildCreateInput",
    "ChildParentFkeyInput": "ChildParentFkeyInput",
    "ChildParentFkeyInverseInput": "ChildParentFkeyInverseInput",
    "ChildParentFkeyParentCreateInput": "ChildParentFkeyParentCreateInput",
    "ChildPatch": "ChildPatch",
    "ChildrenConnection": "ChildrenConnection",
    "ChildrenEdge": "ChildrenEdge",
    "ChildrenOrderBy": "ChildrenOrderBy",
    "CreateChildInput": "CreateChildInput",
    "CreateChildPayload": "CreateChildPayload",
    "CreateParentInput": "CreateParentInput",
    "CreateParentPayload": "CreateParentPayload",
    "Cursor": "Cursor",
    "DeleteChildByParentIdInput": "DeleteChildByParentIdInput",
    "DeleteChildInput": "DeleteChildInput",
    "DeleteChildPayload": "DeleteChildPayload",
    "DeleteParentByIdInput": "DeleteParentByIdInput",
    "DeleteParentInput": "DeleteParentInput",
    "DeleteParentPayload": "DeleteParentPayload",
    "ID": "ID",
    "Int": "Int",
    "Mutation": "Mutation",
    "Node": "Node",
    "PageInfo": "PageInfo",
    "Parent": "Parent",
    "ParentCondition": "ParentCondition",
    "ParentInput": "ParentInput",
    "ParentNodeIdConnect": "ParentNodeIdConnect",
    "ParentNodeIdDelete": "ParentNodeIdDelete",
    "ParentOnChildForChildParentFkeyNodeIdUpdate": "ParentOnChildForChildParentFkeyNodeIdUpdate",
    "ParentOnChildForChildParentFkeyUsingParentPkeyUpdate": "ParentOnChildForChildParentFkeyUsingParentPkeyUpdate",
    "ParentParentPkeyConnect": "ParentParentPkeyConnect",
    "ParentParentPkeyDelete": "ParentParentPkeyDelete",
    "ParentPatch": "ParentPatch",
    "ParentsConnection": "ParentsConnection",
    "ParentsEdge": "ParentsEdge",
    "ParentsOrderBy": "ParentsOrderBy",
    "Query": "Query",
    "String": "String",
    "UpdateChildByParentIdInput": "UpdateChildByParentIdInput",
    "UpdateChildInput": "UpdateChildInput",
    "UpdateChildPayload": "UpdateChildPayload",
    "UpdateParentByIdInput": "UpdateParentByIdInput",
    "UpdateParentInput": "UpdateParentInput",
    "UpdateParentPayload": "UpdateParentPayload",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
    "updateChildOnChildForChildParentFkeyPatch": "updateChildOnChildForChildParentFkeyPatch",
    "updateParentOnChildForChildParentFkeyPatch": "updateParentOnChildForChildParentFkeyPatch",
  },
  "astNode": undefined,
  "description": undefined,
  "extensionASTNodes": undefined,
  "extensions": undefined,
}
`;

exports[`still plural when one-to-one if nestedMutationsOldUniqueFields is enabled 1`] = `
GraphQLSchema {
  "__validationErrors": Array [],
  "_directives": Array [
    "@include",
    "@skip",
    "@deprecated",
    "@specifiedBy",
  ],
  "_implementationsMap": Object {
    "Node": Object {
      "interfaces": Array [],
      "objects": Array [
        "Query",
        "Child",
        "Parent",
      ],
    },
  },
  "_mutationType": "Mutation",
  "_queryType": "Query",
  "_subTypeMap": Object {},
  "_subscriptionType": undefined,
  "_typeMap": Object {
    "Boolean": "Boolean",
    "Child": "Child",
    "ChildChildPkeyConnect": "ChildChildPkeyConnect",
    "ChildChildPkeyDelete": "ChildChildPkeyDelete",
    "ChildCondition": "ChildCondition",
    "ChildInput": "ChildInput",
    "ChildNodeIdConnect": "ChildNodeIdConnect",
    "ChildNodeIdDelete": "ChildNodeIdDelete",
    "ChildOnChildForChildParentFkeyNodeIdUpdate": "ChildOnChildForChildParentFkeyNodeIdUpdate",
    "ChildOnChildForChildParentFkeyUsingChildPkeyUpdate": "ChildOnChildForChildParentFkeyUsingChildPkeyUpdate",
    "ChildParentFkeyChildCreateInput": "ChildParentFkeyChildCreateInput",
    "ChildParentFkeyInput": "ChildParentFkeyInput",
    "ChildParentFkeyInverseInput": "ChildParentFkeyInverseInput",
    "ChildParentFkeyParentCreateInput": "ChildParentFkeyParentCreateInput",
    "ChildPatch": "ChildPatch",
    "ChildrenConnection": "ChildrenConnection",
    "ChildrenEdge": "ChildrenEdge",
    "ChildrenOrderBy": "ChildrenOrderBy",
    "CreateChildInput": "CreateChildInput",
    "CreateChildPayload": "CreateChildPayload",
    "CreateParentInput": "CreateParentInput",
    "CreateParentPayload": "CreateParentPayload",
    "Cursor": "Cursor",
    "DeleteChildByParentIdInput": "DeleteChildByParentIdInput",
    "DeleteChildInput": "DeleteChildInput",
    "DeleteChildPayload": "DeleteChildPayload",
    "DeleteParentByIdInput": "DeleteParentByIdInput",
    "DeleteParentInput": "DeleteParentInput",
    "DeleteParentPayload": "DeleteParentPayload",
    "ID": "ID",
    "Int": "Int",
    "Mutation": "Mutation",
    "Node": "Node",
    "PageInfo": "PageInfo",
    "Parent": "Parent",
    "ParentCondition": "ParentCondition",
    "ParentInput": "ParentInput",
    "ParentNodeIdConnect": "ParentNodeIdConnect",
    "ParentNodeIdDelete": "ParentNodeIdDelete",
    "ParentOnChildForChildParentFkeyNodeIdUpdate": "ParentOnChildForChildParentFkeyNodeIdUpdate",
    "ParentOnChildForChildParentFkeyUsingParentPkeyUpdate": "ParentOnChildForChildParentFkeyUsingParentPkeyUpdate",
    "ParentParentPkeyConnect": "ParentParentPkeyConnect",
    "ParentParentPkeyDelete": "ParentParentPkeyDelete",
    "ParentPatch": "ParentPatch",
    "ParentsConnection": "ParentsConnection",
    "ParentsEdge": "ParentsEdge",
    "ParentsOrderBy": "ParentsOrderBy",
    "Query": "Query",
    "String": "String",
    "UpdateChildByParentIdInput": "UpdateChildByParentIdInput",
    "UpdateChildInput": "UpdateChildInput",
    "UpdateChildPayload": "UpdateChildPayload",
    "UpdateParentByIdInput": "UpdateParentByIdInput",
    "UpdateParentInput": "UpdateParentInput",
    "UpdateParentPayload": "UpdateParentPayload",
    "__Directive": "__Directive",
    "__DirectiveLocation": "__DirectiveLocation",
    "__EnumValue": "__EnumValue",
    "__Field": "__Field",
    "__InputValue": "__InputValue",
    "__Schema": "__Schema",
    "__Type": "__Type",
    "__TypeKind": "__TypeKind",
    "updateChildOnChildForChildParentFkeyPatch": "updateChildOnChildForChildParentFkeyPatch",
    "updateParentOnChildForChildParentFkeyPatch": "updateParentOnChildForChildParentFkeyPatch",
  },
  "astNode": undefined,
  "description": undefined,
  "extensionASTNodes": undefined,
  "extensions": undefined,
}
`;
